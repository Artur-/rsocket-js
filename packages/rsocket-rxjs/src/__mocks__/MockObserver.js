/** Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

import emptyFunction from 'fbjs/lib/emptyFunction';

import {Observer, PartialObserver} from 'rxjs';

export function genMockObserver<T>(
  partialObserver: PartialObserver<T> = {},
): Observer<T> {
  const observer: Observer = {
    complete: jest.fn(partialObserver.complete || emptyFunction),
    error: jest.fn(partialObserver.error || emptyFunction),
    next: jest.fn(partialObserver.next || emptyFunction),
  };
  observer.mockClear = () => {
    observer.complete.mockClear();
    observer.error.mockClear();
    observer.next.mockClear();
  };
  return observer;
}
