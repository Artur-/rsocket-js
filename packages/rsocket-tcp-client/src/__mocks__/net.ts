const EventEmitter = require("events");

export class MockSocket extends EventEmitter {
  end = jest.fn(() => {
    // 'end' is only emitted when a FIN packet is received
    this.emit("close");
  });

  write = jest.fn();

  mock = {
    close: () => {
      this.emit("close");
    },
    connect: () => {
      this.emit("connect");
    },
    data: (data) => {
      this.emit("data", data);
    },
    error: (error) => {
      this.emit("error", error);
    },
  };

  destroy() {}
}

export const net = {
  connect: jest.fn(() => {
    const socket = new MockSocket();
    net.socket = socket; // for easy accessibility in tests
    return socket;
  }),
  socket: null,
};
