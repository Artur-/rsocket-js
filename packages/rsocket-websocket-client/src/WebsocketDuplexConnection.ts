import {
  Closeable,
  Deferred,
  Demultiplexer,
  Deserializer,
  DuplexConnection,
  Frame,
  FrameHandler,
  Multiplexer,
  Outbound,
  serializeFrame,
} from "@rsocket/rsocket-core";
import WebSocket, { CloseEvent, ErrorEvent } from "ws";

export class WebsocketDuplexConnection
  extends Deferred
  implements DuplexConnection, Outbound {
  readonly multiplexerDemultiplexer: Multiplexer &
    Demultiplexer &
    FrameHandler &
    Closeable;

  constructor(
    private websocket: WebSocket,
    private deserializer: Deserializer,
    multiplexerDemultiplexerFactory: (
      outbound: Outbound
    ) => Multiplexer & Demultiplexer & FrameHandler & Closeable
  ) {
    super();

    websocket.addEventListener("close", this.handleClosed.bind(this));
    websocket.addEventListener("error", this.handleError.bind(this));
    websocket.addEventListener("message", this.handleMessage.bind(this));

    this.multiplexerDemultiplexer = multiplexerDemultiplexerFactory(this);
  }

  get availability(): number {
    return this.done ? 0 : 1;
  }

  close(error?: Error) {
    if (this.done) {
      super.close(error);
      return;
    }

    this.websocket.removeEventListener("close", this.handleClosed.bind(this));
    this.websocket.removeEventListener("error", this.handleError.bind(this));
    this.websocket.removeEventListener(
      "message",
      this.handleMessage.bind(this)
    );

    this.websocket.close();

    delete this.websocket;

    this.multiplexerDemultiplexer.close(error);

    super.close(error);
  }

  send(frame: Frame): void {
    if (this.done) {
      return;
    }

    const buffer = serializeFrame(frame);

    this.websocket.send(buffer);
  }

  private handleClosed(e: CloseEvent): void {
    this.close(
      new Error(
        e.reason || "WebsocketDuplexConnection: Socket closed unexpectedly."
      )
    );
  }

  private handleError(e: ErrorEvent): void {
    this.close(e.error);
  }

  private handleMessage = (message: MessageEvent): void => {
    try {
      const buffer = Buffer.from(message.data);
      const frame = this.deserializer.deserializeFrame(buffer);

      this.multiplexerDemultiplexer.handle(frame);
    } catch (error) {
      this.close(error);
    }
  };
}
