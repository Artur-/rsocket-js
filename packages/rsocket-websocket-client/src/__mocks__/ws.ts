import { CloseEvent, ErrorEvent, MessageEvent } from "ws";

const EventEmitter = require("events");

export class MockSocket extends EventEmitter {
  send = jest.fn();

  close = jest.fn();

  addEventListener = (name, handler) => {
    this.on(name, handler);
  };

  removeEventListener = (name, handler) => {
    this.on(name, handler);
  };

  mock = {
    close: (event: Partial<CloseEvent>) => {
      this.emit("close", event);
    },
    open: () => {
      this.emit("connect");
    },
    message: (message: Partial<MessageEvent>) => {
      this.emit("message", message);
    },
    error: (event: Partial<ErrorEvent>) => {
      this.emit("error", event);
    },
  };
}
