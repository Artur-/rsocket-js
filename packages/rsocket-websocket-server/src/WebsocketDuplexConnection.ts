import {
  ClientServerInputMultiplexerDemultiplexer,
  Demultiplexer,
  deserializeFrame,
  DuplexConnection,
  Frame,
  Multiplexer,
  serializeFrame,
  StreamIdGenerator,
} from "@rsocket/rsocket-core";
import { Duplex } from "stream";

export class WebsocketDuplexConnection
  extends ClientServerInputMultiplexerDemultiplexer
  implements DuplexConnection {
  constructor(
    private websocketDuplex: Duplex,
    private connectionAcceptor: (
      frame: Frame,
      connection: DuplexConnection
    ) => Promise<void>
  ) {
    super(StreamIdGenerator.create(0));

    websocketDuplex.on("close", this.handleClosed.bind(this));
    websocketDuplex.on("error", this.handleError.bind(this));
    websocketDuplex.once("data", this.handleFirst.bind(this));
  }

  get multiplexer(): Multiplexer {
    return this;
  }

  get demultiplexer(): Demultiplexer {
    return this;
  }

  get availability(): number {
    return this.websocketDuplex.destroyed ? 0 : 1;
  }

  close(error?: Error) {
    if (this.done) {
      super.close(error);
      return;
    }

    this.websocketDuplex.removeAllListeners();
    this.websocketDuplex.end();

    delete this.websocketDuplex;

    super.close(error);
  }

  send(frame: Frame): void {
    if (this.done) {
      return;
    }

    //   if (__DEV__) {
    //     if (this._options.debug) {
    //       console.log(printFrame(frame));
    //     }
    //   }
    const buffer = /* this._options.lengthPrefixedFrames
          ? serializeFrameWithLength(frame, this._encoders)
          :*/ serializeFrame(
      frame
    );
    // if (!this._socket) {
    //   throw new Error(
    //     "RSocketWebSocketClient: Cannot send frame, not connected."
    //   );
    // }
    this.websocketDuplex.write(buffer);
  }

  private handleClosed(e: CloseEvent): void {
    this.close(
      new Error(
        e.reason || "WebsocketDuplexConnection: Socket closed unexpectedly."
      )
    );
  }

  private handleError(e: ErrorEvent): void {
    this.close(e.error);
  }

  private async handleFirst(buffer: Buffer): Promise<void> {
    try {
      this.websocketDuplex.pause();
      this.websocketDuplex.on("data", this.handleMessage.bind(this));
      const frame = /* this._options.lengthPrefixedFrames
          ? deserializeFrameWithLength(buffer, this._encoders)
          :  */ deserializeFrame(
        buffer
      );
      await this.connectionAcceptor(frame, this);
      this.websocketDuplex.resume();
    } catch (error) {
      this.close(error);
    }
  }

  private handleMessage(buffer: Buffer): void {
    try {
      const frame = /* this._options.lengthPrefixedFrames
          ? deserializeFrameWithLength(buffer, this._encoders)
          :  */ deserializeFrame(
        buffer
      );
      // if (__DEV__) {
      //   if (this._options.debug) {
      //     console.log(printFrame(frame));
      //   }
      // }
      this.handle(frame);
    } catch (error) {
      this.close(error);
    }
  }
}
