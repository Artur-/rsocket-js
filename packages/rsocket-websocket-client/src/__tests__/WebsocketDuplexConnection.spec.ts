import WebSocket from "ws";
import { mock } from "jest-mock-extended";
import {
  Deserializer,
  Flags,
  FrameHandler,
  FrameTypes,
  serializeFrame,
  SetupFrame,
} from "@rsocket/rsocket-core";
import { WebsocketDuplexConnection } from "../WebsocketDuplexConnection";
import { MockSocket } from "../__mocks__/ws";

const deserializer = mock<Deserializer>();

describe("WebsocketDuplexConnection", function () {
  describe("when closed", () => {
    it("removes listeners from the underlying socket event emitter", () => {
      // arrange
      const socketStub = mock<WebSocket>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );

      connection.close();

      expect(socketStub.removeEventListener).toBeCalledWith(
        "close",
        expect.any(Function)
      );
      expect(socketStub.removeEventListener).toBeCalledWith(
        "error",
        expect.any(Function)
      );
      expect(socketStub.removeEventListener).toBeCalledWith(
        "message",
        expect.any(Function)
      );
    });

    it("cleans up the socket resource when closed without an error", () => {
      // arrange
      const socketStub = mock<WebSocket>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );

      connection.close();

      expect(socketStub.close).toBeCalledWith();
    });

    it("cleans up the socket resource when closed with an error", () => {
      // arrange
      const socketStub = mock<WebSocket>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );

      const error = new Error();
      connection.close(error);

      expect(socketStub.close).toBeCalledWith();
    });

    it("calls onClose", () => {
      const socketStub = mock<WebSocket>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );
      const onCloseCallback = jest.fn();

      connection.onClose(onCloseCallback);
      connection.close();

      expect(onCloseCallback).toBeCalledTimes(1);
      expect(onCloseCallback).toBeCalledWith();
    });

    it("calls onClose when closed with an error", () => {
      const socketStub = mock<WebSocket>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );
      const onCloseCallback = jest.fn();
      const error = new Error();

      connection.onClose(onCloseCallback);
      connection.close(error);

      expect(onCloseCallback).toBeCalledTimes(1);
      expect(onCloseCallback).toBeCalledWith(error);
    });

    it("subsequent calls to close result in only a single invocation of onClose", () => {
      const socketStub = mock<WebSocket>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );
      const onCloseCallback = jest.fn();
      const error = new Error();
      connection.onClose(onCloseCallback);
      connection.close(error);
      connection.close(error);

      expect(onCloseCallback).toBeCalledTimes(1);
      expect(onCloseCallback).toBeCalledWith(error);
    });

    it("the onClose callback is called with an error when the socket is closed unexpectedly", () => {
      const socket = (new MockSocket() as unknown) as WebSocket;
      const connection = new WebsocketDuplexConnection(socket, deserializer);
      const onCloseCallback = jest.fn();

      connection.onClose(onCloseCallback);
      ((socket as unknown) as MockSocket).mock.close({});

      expect(onCloseCallback).toBeCalledTimes(1);
      expect(onCloseCallback).toHaveBeenCalledWith(
        new Error("WebsocketDuplexConnection: Socket closed unexpectedly.")
      );
    });

    it("the onClose callback is called with an error when the socket is closed with an error", () => {
      const socket = (new MockSocket() as unknown) as WebSocket;
      const connection = new WebsocketDuplexConnection(socket, deserializer);
      const onCloseCallback = jest.fn();
      const expectedError = new Error(
        "WebsocketDuplexConnection: Test error 1"
      );

      connection.onClose(onCloseCallback);
      ((socket as unknown) as MockSocket).mock.error({ error: expectedError });

      expect(onCloseCallback).toBeCalledTimes(1);
      expect(onCloseCallback).toHaveBeenCalledWith(expectedError);
    });
  });

  describe("handle()", () => {
    it("throws if called twice", () => {
      // arrange
      const socketStub = mock<WebSocket>();
      const frameHandlerStub = mock<FrameHandler>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );

      // assert
      expect(
        connection.connectionInbound.bind(
          connection,
          frameHandlerStub.handle.bind(frameHandlerStub)
        )
      ).not.toThrow();
      expect(
        connection.connectionInbound.bind(
          connection,
          frameHandlerStub.handle.bind(frameHandlerStub)
        )
      ).toThrow("Connection frame handler has already been installed");
      expect(
        connection.handleRequestStream.bind(
          connection,
          frameHandlerStub.handle.bind(frameHandlerStub)
        )
      ).not.toThrow();
      expect(
        connection.handleRequestStream.bind(
          connection,
          frameHandlerStub.handle.bind(frameHandlerStub)
        )
      ).toThrow("Stream handler has already been installed");
    });
  });

  describe("send()", () => {
    const setupFrame = {
      type: FrameTypes.SETUP,
      dataMimeType: "application/octet-stream",
      metadataMimeType: "application/octet-stream",
      keepAlive: 60000,
      lifetime: 300000,
      metadata: Buffer.from("hello world"),
      data: Buffer.from("hello world"),
      resumeToken: null,
      streamId: 0,
      majorVersion: 1,
      minorVersion: 0,
      flags: Flags.METADATA,
    } as SetupFrame;

    it("serializes and writes the given frame to the underlying socket", () => {
      // arrange
      const socketStub = mock<WebSocket>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );

      // act
      connection.send(setupFrame);

      // assert
      expect(socketStub.send).toBeCalledWith(expect.any(Buffer));
    });

    it("does not write the given frame to the underlying socket when close was previously called", () => {
      // arrange
      const socketStub = mock<WebSocket>();
      const connection = new WebsocketDuplexConnection(
        socketStub,
        deserializer
      );

      // act
      connection.close();
      connection.send(setupFrame);

      // assert
      expect(socketStub.send).toBeCalledTimes(0);
    });
  });

  describe("when receiving data", () => {
    const setupFrame: SetupFrame = {
      type: FrameTypes.SETUP,
      dataMimeType: "application/octet-stream",
      metadataMimeType: "application/octet-stream",
      keepAlive: 60000,
      lifetime: 300000,
      metadata: Buffer.from("hello world"),
      data: Buffer.from("hello world"),
      resumeToken: null,
      streamId: 0,
      majorVersion: 1,
      minorVersion: 0,
      flags: Flags.METADATA,
    };

    describe("when buffer contains a single frame", () => {
      it("deserializes received frames and calls the configured handler", () => {
        // arrange
        const handler = mock<FrameHandler>();
        const socketStub = (new MockSocket() as unknown) as WebSocket;
        const connection = new WebsocketDuplexConnection(
          socketStub,
          new Deserializer()
        );

        // act
        connection.connectionInbound(handler.handle.bind(handler));
        ((socketStub as unknown) as MockSocket).mock.message({
          data: serializeFrame(setupFrame),
        });

        // assert
        expect(handler.handle).toBeCalledTimes(1);

        const [call0] = handler.handle.mock.calls;
        const [arg0] = call0;
        expect(arg0).toMatchSnapshot();
      });
    });

    describe("causes an error", () => {
      it("the connection is closed", () => {
        // arrange
        const handler = mock<FrameHandler>();
        const socketStub = (new MockSocket() as unknown) as WebSocket;
        const deserializerStub = mock<Deserializer>();
        const connection = new WebsocketDuplexConnection(
          (socketStub as unknown) as WebSocket,
          deserializerStub
        );
        deserializerStub.deserializeFrame.mockImplementation(() => {
          throw new Error("Mock error");
        });
        const onCloseCallback = jest.fn();
        const data = Buffer.from([]).toString();

        // act
        connection.connectionInbound(handler.handle.bind(handler));
        connection.onClose(onCloseCallback);
        ((socketStub as unknown) as MockSocket).mock.message({ data });

        // assert
        expect(onCloseCallback).toBeCalledTimes(1);
        expect(onCloseCallback).toBeCalledWith(expect.any(Error));
      });
    });
  });
});
