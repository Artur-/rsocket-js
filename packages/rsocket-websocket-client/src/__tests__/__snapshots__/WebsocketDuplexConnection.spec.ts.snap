// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebsocketDuplexConnection when receiving data when buffer contains a single frame deserializes received frames and calls the configured handler 1`] = `
Object {
  "data": Object {
    "data": Array [
      104,
      101,
      108,
      108,
      111,
      32,
      119,
      111,
      114,
      108,
      100,
    ],
    "type": "Buffer",
  },
  "dataMimeType": "application/octet-stream",
  "flags": 256,
  "keepAlive": 60000,
  "lifetime": 300000,
  "majorVersion": 1,
  "metadata": Object {
    "data": Array [
      104,
      101,
      108,
      108,
      111,
      32,
      119,
      111,
      114,
      108,
      100,
    ],
    "type": "Buffer",
  },
  "metadataMimeType": "application/octet-stream",
  "minorVersion": 0,
  "resumeToken": null,
  "streamId": 0,
  "type": 1,
}
`;
